misc/amia_paper_2024/stats.py

```diff
@@ -0,0 +1,77 @@
+"""AMIA paper 2024: Statistics
+
+Resources
+- GH issue: https://github.com/jhu-bids/TermHub/issues/631
+- Paper: https://livejohnshopkins-my.sharepoint.com/:w:/g/personal/sgold15_jh_edu/EXbuxwBsb0pPgFfgPYjR6qgBlJ1gmpqsMPYJ3vn6Y_HnYw?e=4%3ApEDbMe&fromShare=true&at=9&CID=d4663c40-1ebe-db59-8cf5-2bc091d45692"""
+import os
+import sys
+from datetime import datetime
+from pathlib import Path
+from typing import Dict, List
+
+import pandas as pd
+
+THIS_DIR = Path(os.path.dirname(__file__))
+PROJECT_ROOT = THIS_DIR.parent.parent
+sys.path.insert(0, str(PROJECT_ROOT))
+from backend.routes.db import usage_query
+
+
+OUTDIR = THIS_DIR / 'output'
+CSV_PATH = OUTDIR / 'usage.csv'
+
+
+def setup():
+    """Setup any environmental prerequisites"""
+    if not OUTDIR.exists():
+        OUTDIR.mkdir()
+
+
+def get_data() -> pd.DataFrame:
+    """Get data to be analyzed"""
+    if not os.path.exists(CSV_PATH):
+        # JSON_FILE = '/Users/joeflack4/Desktop/vs-hub-usage-report.json'
+        # with open(JSON_FILE, 'r') as file:
+        #     d = json.load(file)
+        # Fetch
+        data: List[Dict] = usage_query()
+        df = pd.DataFrame(data)
+        # Save
+        t0 = datetime.now()
+        df.to_csv(CSV_PATH, index=False)
+        print(f'Wrote {len(df)} records to CSV in n seconds: {(datetime.now() - t0).seconds}')
+    else:
+        # Read
+        t0 = datetime.now()
+        df = pd.read_csv(CSV_PATH)
+        print(f'Read {len(df)} records from CSV in n seconds: {(datetime.now() - t0).seconds}')
+    return df
+
+
+def preprocess(df: pd.DataFrame, verbose=True) -> pd.DataFrame:
+    """Preprocess data"""
+    # Remove developers
+    developer_ips = {'Siggie': '216.164.48.98', 'Joe': '174.99.54.40'}
+    df2 = df[~df['client'].isin(list(developer_ips.values()))]
+    if verbose:
+        print('Filtered out n records based on IP addresses of each dev:')
+        for dev, ip in developer_ips.items():
+            df_i = df[df['client'] == ip]
+            print(f'  - {dev}: {len(df_i)}')
+    # Remove records with no 'API call group'
+    # - These are cases from before we added this feature, or where we called the backend directly.
+    df2 = df2[~df2['api_call_group_id'].isna()]
```

Comment:
## Null `api_call_group_id` preprocessing
I removed these. I didn't count how many rows, but there weren't many. Like hundreds out of the total ~52k.

Reply:
Now there are none of them after my new query. It seems these only exist in the`apiruns_grouped` tablee